File: biblioteca.hs
import System.Directory
 
cadastra:: String -> String -> String -> IO()
cadastra nome autor editora = do
 appendFile "titulos.txt" (nome++ ",")
 appendFile "autores.txt" (autor ++ ",")
 appendFile "editora.txt" (editora ++ ",")
 

listagemLivros = do
   nomeLivro <- readFile "titulos.txt"
   if(length nomeLivro == 0) then do
      putStrLn "Livro Inexistente ou não cadastrado"
   else do
      (a:as) <- readFile ("titulos.txt")
      (b:bs) <- readFile ("autores.txt")
      (c:cs) <- readFile ("editora.txt")
 
      let (nomeLivro:outrosLivros) = listaLivros(a:as) "" []
      let (autor:outroAutores) = listaLivros (b:bs) "" []
      let (editora:outraEditora) = listaLivros(c:cs) "" []
      print (nomeLivro:outrosLivros)
      print (autor:outroAutores)
      print (editora:outraEditora)
 
      imprime (nomeLivro:outrosLivros) (autor:outroAutores) (editora:outraEditora)
     
 
imprime:: [String] -> [String] -> [String] -> IO()
imprime [""] [""] [""]  = putStrLn ("Listagem Encerrada")
imprime [] [] []  = putStrLn ("Listagem Encerrada")
imprime (livro:outrosLivros) (autor:outroAutores) (editora:outraEditora) = do
      putStrLn ("Nome Livro: " ++ livro)
      putStrLn ("Autor: " ++ autor)
      putStrLn ("Editora: " ++ editora)
      imprime outrosLivros outroAutores outraEditora
      
-- funcao que gera uma lista de todos os elementos do arquivo
listaLivros:: String -> String -> [String] -> [String]
listaLivros [] nome [] = [nome]
listaLivros [] nome lista = lista ++ [nome]
listaLivros (x:xs) i lista = do
     if([x] == ",") then do
         listaLivros xs "" (lista ++ [i])
     else if( [x] /= "," && [x] /= "") then do
         listaLivros xs (i++[x]) lista
     else
         lista ++ [i]
------------------------------------------------------------------
-- escreve arquivo dos livros alugados
escreveFileAlugados :: String -> IO()
escreveFileAlugados alugado = do
  appendFile "alugados.txt" (alugado++ ",")
 
-- atualiza (salva nomes na lista de livros alugados)
atualiza :: String -> [String] -> [String]
atualiza nome [""] = []
atualiza nome (x:alugados) = do
         if(nome /= x) then do
             [x] ++  atualiza nome alugados
          else
             atualiza nome alugados
             
-- verifica se o livro ja esta alugado
verificaAlugado :: String -> [String] -> Bool
verificaAlugado titulo [] = True
verificaAlugado titulo [""] = True
verificaAlugado titulo (x:xs)= if(x == titulo) then True
                               else verificaAlugado titulo xs
                               
-- reescreve o arquivo de alugado em branco
reescreve :: IO()
reescreve = writeFile "alugados.txt" ("")

-- reescreve o arquivo com a nova lista de alugados 
reescreveArquivo ::[String] -> IO()
reescreveArquivo [] = putStrLn ""
reescreveArquivo[""] = putStrLn ""
reescreveArquivo (x:xs) = do
      escreveFileAlugados x
      reescreveArquivo xs
 
-----------------------------
main:: IO()
main = do
     print "BIBLIOTECA UFCG"
     putStrLn("Escolha uma das opções seguintes:")
     putStrLn("1 - Cadastra Livro ")
     putStrLn("2 - Lista Livros ")
     putStrLn("3 - Aluga Livros ")
     putStrLn("4 - Devolução")
     
     op <- getLine
     if (op == "1") then do
         putStrLn "CADASTRO DE LIVRO - Digite o nome do livro que deseja cadastrar:"
         nome <- getLine
         autor<- getLine
         editora <- getLine
         cadastra nome autor editora
         
     else if( op == "2") then do
         putStrLn "LISTAGEM DE LIVROS"
         putStrLn "deseja listar todos os livros da biblioteca? S/N"
         resposta <- getLine
         if(resposta == "S") then do
          listagemLivros
         else do
           putStrLn "Digite outra opção:"
     else if( op=="3") then do
     
         putStrLn "ALUGUEL DE LIVROS - Digite o nome do livro que deseja alugar:"
         (x:xs) <- readFile "alugados.txt"
         let livros = listaLivros (x:xs) "" []
         alugado <- getLine
         
         if(verificaAlugado alugado livros) then do 
          putStrLn "Livro não disponível para aluguel"
         else do
           escreveFileAlugados alugado
           putStrLn "Livro alugado com sucesso!"
     else do
            devolvido <- getLine
            (x:xs) <- readFile "alugados.txt"
            let livros = listaLivros (x:xs) "" []
            let result = atualiza devolvido livros
            print result
            a <- removeFile "titulos.txt"
            reescreve
            reescreveArquivo result
    
     main

File: Prolog/biblioteca.pl
:- initialization (main).

:- dynamic 
	livro(_,_,_,_),
	alugado(_,_).

runOption(1) :- 
	write("CADASTRO DE LIVRO - Digite o nome do livro que deseja cadastrar:"),nl,
	read(Titulo),
	write("Informe o Autor do Livro: "),nl,
	read(Autor), 
	write("Informe a Editora do Livro: "), nl,
	read(Editora),
	write("Informe o Ano do Livro: "), nl,
	read(Ano),
	assertz(livro(Titulo, Autor, Editora, Ano)),
	write("Livro cadastrado com sucesso.").

runOption(2) :-
	write("LISTAGEM DE LIVROS"), nl,
	findall(A,livro(A,_,_,_),Lista),
	listagemLivro(Lista).
	

runOption(3) :- 
write("ALUGUEL DE LIVROS - Digite o nome do livro que deseja alugar: "), nl,
	read(Titulo),
	write("Digite o autor do livro: "), nl,
	read(Autor),
	call(livro(Titulo,Autor,_,_)),!,
	aluga(Titulo, Autor);
	write("Livro nao cadastrado.").


runOption(4):-
write("LISTAGEM DE LIVROS ALUGADOS"), nl,
	findall(Livro,alugado(Livro, Autor),Lista),
	listagemAlugados(Lista).

runOption(5):- 
	write("DEVOLUÇAO DE LIVROS - Digite o titulo do livro que deseja devolver: "), nl,
	read(Livro),
	write("Digite o autor do livro: "), nl,
	read(Autor),
	devolve(Livro, Autor).

runOption(6):- write("BUSCA DE LIVROS CADASTRADOS"),nl,
	write("Digite o titulo do Livro: "), nl,
	read(Titulo), 
	write("Digite o Autor: "),nl,
	read(Autor),
	write("Digite a Editora: "),nl,
	read(Editora),
	write("Digite o Ano: "),nl,
	read(Ano),
	call(livro(Titulo,Autor,Editora,Ano)),!,
	write("Livro cadastrado.");
	write("Livro nao cadastrado.").

runOption(7):- write("BUSCA DE LIVROS ALUGADOS"),nl,
	write("Digite o titulo do livro: "), nl,
	read(Titulo),
	write("Digite o autor do livro: "), nl,
	read(Autor),
	buscaAlugado(Titulo,Autor).

runOption(8):- halt(0).
runOption(_):- write("opcao invalida"),main.

%------------------ devolve livro alugado--------------------------
devolve(Titulo, Autor):- 	
	call(alugado(Titulo, Autor)),!,
	retract(alugado(Titulo, Autor)),
	write("Livro devolvido com sucesso!");
	write("Livro nao devolvido.").
	
%---------------- lista livro alugado -------------------------------------	
listagemAlugados([]):- main.
listagemAlugados([HEAD|TAIL]) :-
	write("Titulo: "),
	write(HEAD), nl,
	alugado(HEAD,Autor),
	write("Autor: "),
	write(Autor),nl,
	write("Status: ALUGADO"),nl,
	listagemAlugados(TAIL).	

%-------------------------busca alugado---------------------------
buscaAlugado(Titulo,Autor):- call(alugado(Titulo,Autor)),!,
	write("Livro ja esta alugado.");
	write("Livro disponivel para aluguel").
	
%--------------- lista livros alugados ----------------------------------	
listagemLivro([]):- main.
	listagemLivro([HEAD|TAIL]):- nl, write("Titulo: "),
	write(HEAD), nl,
	livro(HEAD,Autor,Editora, Ano), 
	write("Autor:"),
	write(Autor), nl,
	write("Editora: "),
	write(Editora),nl,
	write("Ano: "),
	write(Ano), nl,
	listagemLivro(TAIL).	
	
%----------- aluga livro -----------------------------------------------	
aluga(Titulo, Autor):- call(alugado(Titulo, Autor)),!,
	write("Livro indisponivel para aluguel");
	assertz(alugado(Titulo, Autor)),
	write("Livro alugado com sucesso").	

main:-
nl,
write("
___  _ ___  _    _ ____ ___ ____ ____ ____    _  _ ____ ____ ____ 
|__] | |__] |    | |  |  |  |___ |    |__|    |  | |___ |    | __ 
|__] | |__] |___ | |__|  |  |___ |___ |  |    |__| |    |___ |__] 
                                                                                                                               
"), nl,
write("1 - Cadastra Livro "),nl,
write("2 - Lista Livros "),nl,
write("3 - Aluga Livros "),nl,
write("4 - Lista Livros Alugados"),nl,
write("5 - Devolução"),nl,
write("6 - Buscar Livro"),nl,
write("7 - Buscar livro alugado"),nl,
write("8 - Sair"),nl,
write("Por favor, informe a sua escolha: "), nl,
read(Option),
runOption(Option), main.
